src/main/java/org/opentripplanner/netex/mapping/ServiceLinkMapper.java

```diff
@@ -1,246 +1,286 @@
           issueStore.add(
             "ServiceLinkWithoutLineString",
             "Ignore linkSequenceProjection without linestring for: %s",
             linkSequenceProjection
           );
+          return null;
+        }
+        List<Double> coords = linkSequenceProjection.getLineString().getPosList().getValue();
+        if (coords.size() < 4) {
+          issueStore.add(
+            "ServiceLinkGeometryError",
+            "Ignore linkSequenceProjection with invalid linestring, " +
+            "containing fewer than two coordinates for: %s",
+            serviceLink.getId()
+          );
+          return null;
+        } else if (coords.size() % 2 != 0) {
+          issueStore.add(
+            "ServiceLinkGeometryError",
+            "Ignore linkSequenceProjection with invalid linestring, " +
+            "containing odd number of values for coordinates: %s",
+            serviceLink.getId()
+          );
+          return null;
+        }
+
+        Coordinate[] coordinates = new Coordinate[coords.size() / 2];
+        for (int i = 0; i < coords.size(); i += 2) {
+          coordinates[i / 2] = new Coordinate(coords.get(i + 1), coords.get(i));
+        }
+        final LineString geometry = geometryFactory.createLineString(coordinates);
+
+        if (
+          !isValid(
+            geometry,
+            tripPattern.getStop(stopIndex),
+            tripPattern.getStop(stopIndex + 1),
+            serviceLink.getId(),
+            issueStore
+          )
+        ) {
+          return null;
         }
+        return geometry;
       }
     }
+
+    issueStore.add(
+      "ServiceLinkWithoutProjection",
+      "Ignore ServiceLink without projection: %s",
+      serviceLink.getId()
+    );
+    return null;
   }
 
-  private void addStraightLine(
-    ServiceLink serviceLink,
-    MutableInt sequenceCounter,
-    MutableDouble distanceCounter,
-    ReadOnlyHierarchicalMap<String, String> quayIdByStopPointRef,
-    ReadOnlyHierarchicalVersionMapById<Quay> quayById,
-    Collection<ShapePoint> shapePoints,
-    FeedScopedId shapePointIdFromJourneyPatternId
-  ) {
-    String fromPointQuayId = quayIdByStopPointRef.lookup(serviceLink.getFromPointRef().getRef());
-    Quay fromPointQuay = quayById.lookupLastVersionById(fromPointQuayId);
+  /** create a 2-point linestring (a straight line segment) between the two stops */
+  private LineString createSimpleGeometry(StopLocation s0, StopLocation s1) {
+    Coordinate[] coordinates = new Coordinate[] {
+      s0.getCoordinate().asJtsCoordinate(),
+      s1.getCoordinate().asJtsCoordinate(),
+    };
+    CoordinateSequence sequence = new PackedCoordinateSequence.Double(coordinates, 2);
 
-    String toPointQuayId = quayIdByStopPointRef.lookup(serviceLink.getToPointRef().getRef());
-    Quay toPointQuay = quayById.lookupLastVersionById(toPointQuayId);
+    return geometryFactory.createLineString(sequence);
+  }
 
+  private boolean isValid(
```

Comment:
could you document here what validity checks are performed?

Reply:
I've split and separated the different parts of the validation, and given them sensible names